README.md
# Placer

&copy; CETIC 2017 www.cetic.be

Placer is a component of the European Project TANGO (http://tango-project.eu ).

Placer is distributed under a [GNU LGPL License, version 3.0 or above](https://www.gnu.org/licenses/lgpl-3.0.en.html).

## Description

Placer is a model-based tool that, given a model of heterogeneous (or at least multi-core) hardware and a task-based complex software, finds a mapping of the software tasks on the various processing elements of the hardware, together with a routing of the transmissions on the available busses, and provides a schedule for the tasks and transmissions.
The mapping can minimize either the run time, energy consumption, or both in a multi-objective fashion.

## Installation

### Requirements

Placer is an R&D Tool that runs on the JVM. As such it can run on any office machine running a JVM, with say 10Gb of RAM available for the tool.

To use Placer you need to install two pieces of software:
* [jdk1.8](http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html).
* [scala 2.11](https://www.scala-lang.org/download/2.11.8.html).

### Installation and configuration procedure

Just copy the .jar of the release, once with requirements are installed and run it through command line.
In the repository, \data contains two examples: example1.json and smallExample.json and the output that should be produced by Placer: placement_of_example1.json and placment_of_smallExample.json, respectively

## Usage:

Placer is a command-line tool. A documentation of all parameter is available through the command line as well using the –help flag:

```
> java -jar Placer.jar --help

Placer version Beta5

Usage: placer [options]

  --help                   prints this usage text
  --in <file>              the JSon file representing the placement problem
  --out <file>             the file where the JSon representing the placements will be stored
  --discrepancy <value>    the maximal discrepancy to use during the search must be >=0 , lower is faster but incomplete, use 20 for instance (5 if in a hurry). default is MaxInt
  --strategy strategy1,strategy2,...
                           Search strategies to use, in the order they must be searched. The available search stategies are:
                                 TransmissionRouting
                                 TaskPlacementLessBuzyProcFirst
                                 LocalOrBusTransmissionLargestFirstLocalFirst
                                 LocalOrBusTransmissionLongestAdjFirstNonLocalFirst
                                 TaskPlacementFastestImplemPlusLessBuzyProcFirst
                                 SharedImplementationInstances
                                 TaskAndTransmissionStarts
                                 HEFT  (not very good in CP context)
                 when not specified, a default search strategy is used.

  --timeLimit <value>      the maximal run time for Placer in seconds, default is MaxInt, also used for LNS
  --verbose                prints some verbosities
  --license                prints license and stops
  --lns <value>            use LNS, only for single objective goal (minMakespan,minEnergy,...) not for sat or Pareto
  --lnsMaxFails <value>    for LNS: the maximal number of fail per CP search default is 2000
  --lnsTimeLimit <value>   for LNS: the time limit for each LNS search.
  --lnsRelaxProba <value>  for LNS: the probability (in percentage) to maintain an element from one solution to the relaxed one, default is 90
  --lnsNbRelaxations <value>
                           for LNS: the total number of relaxation to try out, default is 500
  --lnsNbRelaxationNoImprove <value>
                           for LNS: the maximal number of consecutive relaxation without improvement, default is 200
  --lnsCarryOnObjForMultiHardware <value>
                           when using multi hardware, should Placer carry on the best values for the objective function from one hardware to the next one?
                                 0 is no;
                                 1 is yes, but if first solution cannot be found, try again without carry on (this is the default)
                                 2 is yes, and without retry
  --lnsUseEarlyStop <value>
                           when using LNS, this option ask the solver to first make a search with all limits (time and fails) divided by ten, and stop if this search was not fruitful. If the search was fruitful, it then proceeds with the normal set limits (time and fails)
  --benchmark <value>      run the tool, reports the run time and obj, and optionally saves the outfile, if an outfile is given.
Note: Placer is a java software so that all options taken by the JVM also apply.
Among them, you should consider the -Xmx and -Xms parameters to grant more memory to Placer:
example: java -Xms4G -Xmx15G -jar Placer.jar --in=...
```



another example of usage:
notice that OscaR is quite verbose during its execution, even if verbosities are disabled.

```
> java -Xms5G -Xmx20G Placer.jar  --in=aquascanDemo1.json
		--out=aquascanOut1.json --timeLimit=10

staticMaxHorizon:291553
maxHorizon:      291553
nbTasks:51
nbTransmissions:62
creating tasks
creating processors
constants about adjacency
creating busses
creating transmissions
registering tasks and transmissions to processors and busses
registering execution constraints per processors and task
INFO: skipping tautological cumulative constraint: temporary storage for processor Core1, summedMaxAmount:146506, min available resource:1048576
INFO: skipping tautological cumulative constraint: temporary storage for processor Core2, summedMaxAmount:146506, min available resource:1048576
INFO: skipping tautological cumulative constraint: temporary storage for processor Core3, summedMaxAmount:146506, min available resource:1048576
INFO: skipping tautological cumulative constraint: temporary storage for processor Core4, summedMaxAmount:146506, min available resource:1048576
computing makeSpan on tasks
redundant bin-packing constraint on usage per bus
computing total energy consumption
posting mapping constraints
posting MustBeUsedConstraint(Core1)
posting MustBeUsedConstraint(Core2)
posting MustBeUsedConstraint(Core3)
posting MustBeUsedConstraint(Core4)
breaking symmetry among List(Core1, Core2, Core3, Core4) by workload
starting search
solution found, makeSpan=125024 energy:0
solution found, makeSpan=88643 energy:0
solution found, makeSpan=88642 energy:0
solution found, makeSpan=88641 energy:0
nNodes: 975297
nFails: 487355
time(ms): 10002
completed: false
timeInTrail: 605
nSols: 4
secondLevels:487897
thirdLevels:4


Process finished with exit code 0
```


ReleaseNotes.txt

beta5 release:
--------------

reworked the constraints and objectives
added a lot of new constraints
added the possibility to specify the search procedure through the command line
improved the LNS heuristics
documented the input and output formats

beta4 release:
------------------
not released at all

beta3 release:
--------------
added shared functions on FPGA: a function is instantiated once (or more) and used by several tasks

beta2 release:
--------------
added manyCore and manyThread
added version tag in Json file format
improved reference document

beta1 release:
---------------
modified the JSon format:
processing element classes are in the mapping problem
improved a lot the search strategy
solved a lot of bugs related to the time model and a bug with underlying constraint
uses latest OscaR.cp model
rewrote the model so that it better scales up
added LNS

alpha6 release:
---------------
modified the JSon input format.
timeUnit and dataUnit have to be declared in the MappingProblem
and fields timeUnitPerBit are renamed to timeUnitPerDataUnit

alpha5 release:
---------------
Speed improvement
uses a modified model for the CP engine where zero values for transmission delays are separated from other ones.
This reduces the size of domains declared for CPIntVars

alpha4 release:
---------------
Speed improvement
uses a better search strategy
also added a command line parameter to increase/decrease optimality


src/main/scala/placer/Main.scala
/*
 * Copyright 2017 CETIC www.cetic.be
 * This is being developed for the TANGO Project: http://tango-project.eu
 *
 * Placer is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Placer is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Placer.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 *
 */

package placer

import java.io.{File, PrintWriter}

import net.liftweb.json._
import placer.algo.{Mapper, MapperConfig, Strategy}
import placer.io.Extractor
import placer.metadata.MappingProblem

import scala.io.Source

case class Config(in: File = new File("."),
                  out: File = null,
                  verbose: Boolean = false,
                  license:Boolean = false,
                  discrepancy:Int = Int.MaxValue,
                  strategies:Option[List[Strategy.Value]] = None,
                  timeLimit:Int = Int.MaxValue,
                  lns:Boolean = false,
                  lnsTimeLimit:Int = Int.MaxValue,
                  lnsMaxFails:Int = 2000,
                  lnsRelaxProba:Int = 90,
                  lnsNbRelaxations:Int = 500,
                  lnsNbRelaxationNoImprove:Int = 200,
                  lnsCarryOnObjForMultiHardware:Int = 1,
                  lnsUseEarlyStop:Boolean = true,
                  benchmark:Boolean = false)

object Main extends App {

  val parser = new scopt.OptionParser[Config]("placer") {
    head("placer", "version Beta5")

    help("help").text("prints this usage text")

    opt[File]("in").required().maxOccurs(1).valueName("<file>").
      action((x, c) => c.copy(in = x)).
      text("the JSon file representing the placement problem")

    opt[File]("out").maxOccurs(1).valueName("<file>").
      action((x, c) => c.copy(out = x)).
      text("the file where the JSon representing the placements will be stored")

    opt[Int]("discrepancy").maxOccurs(1).
      action((x, c) => c.copy(discrepancy = x)).
      text("the maximal discrepancy to use during the search must be >=0 , lower is faster but incomplete, use 20 for instance (5 if in a hurry). default is MaxInt")

    opt[Seq[String]]("strategy").maxOccurs(1).valueName("strategy1,strategy2,...").action( (strategies,c) =>
      c.copy(strategies = Some(strategies.toList.map(decodeStrategy)))).text(
      """|Search strategies to use, in the order they must be searched. The available search stategies are:
                                 TransmissionRouting
                                 TaskPlacementLessBuzyProcFirst
                                 LocalOrBusTransmissionLargestFirstLocalFirst
                                 LocalOrBusTransmissionLongestAdjFirstNonLocalFirst
                                 TaskPlacementFastestImplemPlusLessBuzyProcFirst
                                 SharedImplementationInstances
                                 TaskAndTransmissionStarts
                                 HEFT  (not very good in CP context)
                 when not specified, a default search strategy is used.
        """.stripMargin('|'))

    import Strategy._
    def decodeStrategy(s:String):Strategy.Value = {
      s match{
        case "TransmissionRouting" => TransmissionRouting
        case "TaskPlacementLessBuzyProcFirst" => TaskPlacementLessBuzyProcFirst
        case "LocalOrBusTransmissionLargestFirstLocalFirst" => LocalOrBusTransmissionLargestFirstLocalFirst
        case "LocalOrBusTransmissionLongestAdjFirstNonLocalFirst" => LocalOrBusTransmissionLongestAdjFirstNonLocalFirst
        case "TaskPlacementFastestImplemPlusLessBuzyProcFirst" => TaskPlacementFastestImplemPlusLessBuzyProcFirst
        case "SharedImplementationInstances" => SharedImplementationInstances
        case "TaskAndTransmissionStarts" => TaskAndTransmissionStarts
        case "HEFT" => HEFT
        case _ => throw new Error("unknown search strategy:" + s)
      }
    }

    opt[Int]("timeLimit").maxOccurs(1).
      action((x, c) => c.copy(timeLimit = x)).
      text("the maximal run time for Placer in seconds, default is MaxInt, also used for LNS")

    opt[Unit]("verbose").maxOccurs(1).action((_, c) =>
      c.copy(verbose = true)).text("prints some verbosities")

    opt[Unit]("license").maxOccurs(1).action((_, c) =>
      c.copy(license = true)).text("prints license and stops")

    //LNS stuff

    opt[Boolean]("lns").maxOccurs(1).action((bool,c) =>
      c.copy(lns = bool)).text("use LNS, only for single objective goal (minMakespan,minEnergy,...) not for sat or Pareto")

    opt[Int]("lnsMaxFails").maxOccurs(1).
      action((x, c) => c.copy(lnsMaxFails = x)).
      text("for LNS: the maximal number of fail per CP search default is 2000")

    opt[Int]("lnsTimeLimit").maxOccurs(1).
      action((x, c) => c.copy(lnsTimeLimit = x)).
      text("for LNS: the time limit for each LNS search. ")


    opt[Int]("lnsRelaxProba").maxOccurs(1).
      action((x, c) => c.copy(lnsRelaxProba = x)).
      text("for LNS: the probability (in percentage) to maintain an element from one solution to the relaxed one, default is 90")

    opt[Int]("lnsNbRelaxations").maxOccurs(1).
      action((x, c) => c.copy(lnsNbRelaxations = x)).
      text("for LNS: the total number of relaxation to try out, default is 500")

    opt[Int]("lnsNbRelaxationNoImprove").maxOccurs(1).
      action((x, c) => c.copy(lnsNbRelaxationNoImprove = x)).
      text("for LNS: the maximal number of consecutive relaxation without improvement, default is 200")

    opt[Int]("lnsCarryOnObjForMultiHardware").maxOccurs(1).action((int,c) =>
      c.copy(lnsCarryOnObjForMultiHardware = int)).
      text("""|when using multi hardware, should Placer carry on the best values for the objective function from one hardware to the next one?
                                 0 is no;
                                 1 is yes, but if first solution cannot be found, try again without carry on (this is the default)
                                 2 is yes, and without retry""".stripMargin('|'))

    opt[Boolean]("lnsUseEarlyStop").maxOccurs(1).action((bool,c) =>
      c.copy(lnsUseEarlyStop = bool)).text("when using LNS, this option ask the solver to first make a search with all limits (time and fails) divided by ten, " +
      "and stop if this search was not fruitful. If the search was fruitful, it then proceeds with the normal set limits (time and fails)")

    opt[Boolean]("benchmark").maxOccurs(1).action((bool,c) =>
      c.copy(benchmark = bool)).text("run the tool, reports the run time and obj, and optionally saves the outfile, if an outfile is given. ")

    note("""|Note: Placer is a java software so that all options taken by the JVM also apply.
            |Among them, you should consider the -Xmx and -Xms parameters to grant more memory to Placer:
            |example: java -Xms4G -Xmx15G -jar Placer.jar --in=...""".stripMargin('|'))
  }

  parser.parse(args, Config()) match {
    case None =>
      // arguments are bad, error message will have been displayed
      System.err.println("exiting Placer")
      System.exit(-1)
    case Some(config) if config.license =>
      // do stuff
      println(
        """
          |Copyright 2017 CETIC www.cetic.be
          |This is being developed for the TANGO Project: http://tango-project.eu
          |
          |Placer is free software: you can redistribute it and/or modify
          |it under the terms of the GNU Lesser General Public License as published by
          |the Free Software Foundation, either version 3 of the License, or
          |(at your option) any later version.
          |
          |Placer is distributed in the hope that it will be useful,
          |but WITHOUT ANY WARRANTY; without even the implied warranty of
          |MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
          |GNU Lesser General Public License  for more details.
          |
          |You should have received a copy of the GNU Lesser General Public License along with Placer.
          |If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html""".stripMargin('|'))
      System.exit(0)
    case Some(config) =>
      val verbose = config.verbose

      val problemFile = Source.fromFile(config.in)
      val parsed = parse(problemFile.mkString)
      val problem: MappingProblem = Extractor.extractProblem(parsed,verbose)

      require(config.out != null || config.benchmark, "---out must be specified if not running benchmark mode")

      if (config.verbose) println(problem)

      val startTime = System.nanoTime()
      val mappingSet = Mapper.findMapping(problem,
        MapperConfig(maxDiscrepancy = config.discrepancy,
          timeLimit = config.timeLimit,
          lns = config.lns,
          lnsTimeLimit = config.lnsTimeLimit,
          strategy = config.strategies,
          lnsMaxFails = config.lnsMaxFails,
          lnsRelaxProba = config.lnsRelaxProba,
          lnsNbRelaxations = config.lnsNbRelaxations,
          lnsNbRelaxationNoImprove = config.lnsNbRelaxationNoImprove,
          lnsCarryOnObjForMultiHardware = config.lnsCarryOnObjForMultiHardware,
          lnsUseEarlyStop = config.lnsUseEarlyStop,
          verbose = verbose))

      val endTime = System.nanoTime()

      if (config.verbose) println(mappingSet)

      if(config.benchmark){
        println("runTime: " + ((endTime - startTime)/(1000*1000)) + " ms" + " obj: " + mappingSet.mapping.map(_.objValues.mkString(",")))
        if(config.out != null) {
          val outJSon = mappingSet.toJSon
          new PrintWriter(config.out) {
            val parsed = parse(outJSon)
            write(prettyRender(parsed))
            close()
          }
        }
      }else{
        val outJSon = mappingSet.toJSon
        new PrintWriter(config.out) {
          val parsed = parse(outJSon)
          write(prettyRender(parsed))
          close()
        }
      }
      System.exit(0)
  }
}

